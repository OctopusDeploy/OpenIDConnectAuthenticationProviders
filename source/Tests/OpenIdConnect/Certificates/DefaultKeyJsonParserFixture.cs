using System.IO;
using Assent;
using Newtonsoft.Json;
using NUnit.Framework;
using Octopus.Server.Extensibility.Authentication.OpenIDConnect.Common.Certificates;

namespace Tests.OpenIdConnect.Certificates
{
    [TestFixture]
    public class DefaultKeyJsonParserFixture
    {
        [Test]
        public void ShouldParseSupportedJwks_AzureAD()
        {
            // Source: https://login.microsoftonline.com/common/discovery/v2.0/keys
            var content = LoadJwksContent(nameof(ShouldParseSupportedJwks_AzureAD));

            var result = new DefaultKeyJsonParser().Parse(content);

            this.Assent(FormattedResult(result));
        }

        [Test]
        public void ShouldParseSupportedJwks_Google()
        {
            // Source: https://www.googleapis.com/oauth2/v3/certs
            var content = LoadJwksContent(nameof(ShouldParseSupportedJwks_Google));

            var result = new DefaultKeyJsonParser().Parse(content);

            this.Assent(FormattedResult(result));
        }

        [Test]
        public void ShouldParseSupportedJwks_Okta()
        {
            // Source: https://octopus-dev-roy.okta.com/oauth2/v1/keys
            var content = LoadJwksContent(nameof(ShouldParseSupportedJwks_Okta));

            var result = new DefaultKeyJsonParser().Parse(content);

            this.Assent(FormattedResult(result));
        }

        static string FormattedResult(KeyDetails[] result)
        {
            return JsonConvert.SerializeObject(result, Formatting.Indented, new JsonSerializerSettings {TypeNameHandling = TypeNameHandling.All});
        }

        string LoadJwksContent(string testName)
        {
            var fixtureType = typeof(DefaultKeyJsonParserFixture);
            var resourceName = $"{fixtureType.FullName}.{testName}.jwks.json";
            using (var s = fixtureType.Assembly.GetManifestResourceStream(resourceName))
            using (var sr = new StreamReader(s))
            {
                return sr.ReadToEnd();
            }
        }
    }
}