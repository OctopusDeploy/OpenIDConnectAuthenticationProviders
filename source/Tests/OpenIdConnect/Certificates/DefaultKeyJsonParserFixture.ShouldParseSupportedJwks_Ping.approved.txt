{
  "$type": "Octopus.Server.Extensibility.Authentication.OpenIDConnect.Common.Certificates.KeyDetails[], Octopus.Server.Extensibility.Authentication.OpenIDConnect.Common",
  "$values": [
    {
      "$type": "Octopus.Server.Extensibility.Authentication.OpenIDConnect.Common.Certificates.RsaDetails, Octopus.Server.Extensibility.Authentication.OpenIDConnect.Common",
      "Modulus": "kZ3TEh6j93bWSR1Fpkgam2SiBIohC4_KCeUYd-_uUM6jDQZic5pOEvHWxL-sc3Mg4Zx4c5Qn_Lw5BVPtyY2BSDLiQwTCD8RpGX7MgQllgd9TM64PblyE6HPMgCOqIOxDHq-QJL1IlA80OM8Hlf7Ze-YQPeKxfTBpfuw4KcNxblo3G5pNq8tFDjoBqE4E3mOv8Ay1b0KnZZTfxQ-FqXLEynysRFNGZ-cWyvRRCD_1S8f5BppBJq6lnlSwLg5MvFwLd5vCFx_WShYv-pTy2kUascEVnLdTo0WAZcVpUkFaRL5G_TrQe-pe892hF3EYl2I9uVUXjQTlmlB-x2o5MZGFcw",
      "Exponent": "AQAB",
      "Kid": "da7wFyqmnDaQ1WKEhXZ-A6rYqSA"
    },
    {
      "$type": "Octopus.Server.Extensibility.Authentication.OpenIDConnect.Common.Certificates.RsaDetails, Octopus.Server.Extensibility.Authentication.OpenIDConnect.Common",
      "Modulus": "qxCiLX5H67U4d30Mt3DATMpdyIAxsnTfRPivB0RXA93udke6i22BTHHA16ZkBvPeiQOOFirhTmz5SOVib4BoKKPQ4druxOXK8nw9Rb_sppMkWweuE7MqU7vjOAaoHHfxV4Q259H-uU48t2cvLMgYIu7Bjx4j0Trgx_SxQt5MKbIH0Fm0fzgwAzL3-MZ45QpM7ZQNMtrmh33HucM_fQ7kL3eu9M2l50ArnazqwFwwr8sIRexwG5bK_tqDpMiaU1Wd2DSbG1K3Bz2lM7o8MUipJuM1JrvFZmZ28siJsOv5V6qXxsrtEXv6_FnnGsLt0wbG5fafg1jxBGhbjpM1JHNITw",
      "Exponent": "AQAB",
      "Kid": "P4QmQdazhOEZ6ESp3Ox50SdbHaU"
    },
    {
      "$type": "Octopus.Server.Extensibility.Authentication.OpenIDConnect.Common.Certificates.RsaDetails, Octopus.Server.Extensibility.Authentication.OpenIDConnect.Common",
      "Modulus": "ie0wE0bn1pQ0xEeM5-2_I3aMt_eDI7oyCClCyU1faMeQkzDTAp6FLnqSD99Ryb-dheGObooPWm1JxxKU8gvoNl2k7OVNYcyZUsaBmvCLn1RWsBOpULojgjwdAH9WeKEXzNJVBwnGy_MZaRxC10-DVe36tIzhAzqo_I5WC-LKIb5l1MTbXjMVGltuQyVd-TrLhKunsXXBC7pwIEw2hh7XT9eeg8Hm4vYj1Urj44JsbNxKQ1JHEt4qG3RKAEJAipi9uSE8L8cnxTEHkqC9g1uomghazyNc17XMofTTzy7d0ZavRpIfnUID0YAAQihYBJCJP2pQn-lAfRQgjnBdCDZjRQ",
      "Exponent": "AQAB",
      "Kid": "gvqsO-tegUGtLWrVeFtRDjwCzhE"
    }
  ]
}